feat(api): make fw updates optional when building the hardware controller

vegano1 opened the pull request:
# Overview

Having the firmware on the flex auto-update is great, as it allows the subsystems to stay in sync with the OS build. However, if we are running the hardware controller from CLI we won't know if updates are running which can be confusing. This PR makes firmware updates optional for the purpose of using the hardware controller via other scripts, as well as prints the update progress to stdout so we know if something is updating.

# Test Plan

- [x] Make sure that firmware updates still run on start-up.
- [x] Make sure we can disable firmware updates when running repl with new `OT3_DISABLE_FW_UPDATES` env variable

# Changelog

- add `update_required` optional argument when building the hardware controller which enables/disables auto updates.
- add `OT3_DISABLE_FW_UPDATES` env flag for repl to disable fw updates from CLI.
- print update progress to stdout so we know when updates are running via CLI

# Review requests

- Does this make sense?

# Risk assessment

Low, keeps existing behavior

vegano1 committed:
feat(api): make fw updates optional when building the hardware controller

andySigler reviewed (commented):
i see it's still in draft but this is would be awesome, most especially that `print(progress)` part 👍

vegano1 committed:
format

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -292,8 +293,11 @@ async def build_hardware_controller(
+            async for progress in api_instance.update_firmware():
+                # NOTE: We are printing to stdout so when the hw controller is
+                # built for a CLI, the operator can know when updates are running.
+                print(progress)
```
This is not the appropriate place for this, I think. This is core library code, and is not the place for contextual use case things like this. IMO we should do one of the following:
- Make this a `log.info` and configure the repl to stream logs to stdout before calling `build`, and then reconfigure back afterward
- that's kind of the only thing we can do here. Maybe we could separate this into subfunctions and call them separately instead or something.

Either way, definitely can't have a print here, and it's worth doing elaborate workarounds to avoid.

sfoster1 reviewed (changes requested):
I see the point of the `print`, but I still think we can't have it in the core api code.

vegano1 committed:
use StreamHandler to log firmware update progress to stdout when building the hardware controller.

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -292,8 +294,18 @@ async def build_hardware_controller(
+        if update_firmware:
+            # NOTE: We are logging to stdout so when the hw controller is
+            # built for CLI, the operator can know when updates are running.
+            stdout_handler = logging.StreamHandler()
```
Hold on. when i said "print" what i meant was "emitting directly to stdout". Line 300 and line 308 should live _in the repl_. Library code should never emit to stdout, and only maybe emit to stderr

sfoster1 commented on api/src/opentrons/hardware_control/scripts/repl.py:
```diff
@@ -108,7 +112,9 @@ def stop_server() -> None:
 
 def build_api() -> ThreadManager[HardwareControlAPI]:
     tm = ThreadManager(
```
We can configure that stream handler, with the exact same code as is currently in the `build` function, right here to get the effect we want.

sfoster1 reviewed (changes requested):
Sorry, wasn't specific enough - what I don't like about `print` in library code isn't the use of the literal `print` function, but generally "emitting to stdout". Library code should never emit to stdout and only maybe to stderr. We can keep that exact logging configuration code but move it to `repl.py` and get the thing we want.

vegano1 committed:
move StreamHandler to repl
set the `ot3Controller.initialized` based on wether firmware updates are enabled or not, this way we dont raise `FirmwareUpdateRequired` exception when firmware updates are disabled.

vegano1 commented:
@sfoster1 

Had to fiddle around with getting the right logger, and handling `FirmwareUpdateRequired` exception when updates are disabled, but this should now do what we want.

sfoster1 committed:
couple fixes, add a test

sfoster1 reviewed (approved):
i like it but i did just make changes to it

sfoster1 commented:
Tested on a robot, does what we want!

ahiuchingau reviewed (approved):
YAY looks good to me! 

I think we should add `update_fw: bool = False` to the [hardware controller builder function ](https://github.com/Opentrons/opentrons/blob/edge/hardware-testing/hardware_testing/opentrons_api/helpers_ot3.py#L124)signature in `hardware-testing` to make sure the scripts are using this new feature.

sfoster1 commented:
> YAY looks good to me!
> 
> I think we should add `update_fw: bool = False` to the [hardware controller builder function ](https://github.com/Opentrons/opentrons/blob/edge/hardware-testing/hardware_testing/opentrons_api/helpers_ot3.py#L124)signature in `hardware-testing` to make sure the scripts are using this new feature.

I really don't think we should hehe. This is a one off to prevent some annoying delays during repl usage. anything, and i mean _anything_, systematically run needs updated firmware.
